lib/shopify_app/session/session_repository.rb

```diff
@@ -2,21 +2,23 @@ module ShopifyApp
   class SessionRepository
     class ConfigurationError < StandardError; end
 
+    EXPECTED_METHODS = [:store, :retrieve].freeze
+    EXPECTED_SHOP_STORAGE_METHODS = (EXPECTED_METHODS + [:retrieve_by_shopify_domain]).freeze
+    EXPECTED_USER_STORAGE_METHODS = (EXPECTED_METHODS + [:retrieve_by_shopify_user_id]).freeze
+
     class << self
       def shop_storage=(storage)
         @shop_storage = storage
+        return unless storage
 
-        unless storage.nil? || self.shop_storage.respond_to?(:store) && self.shop_storage.respond_to?(:retrieve)
-          raise ArgumentError, "shop storage must respond to :store and :retrieve"
-        end
+        raise ArgumentError, "shop storage does not have expected interface" unless expected_shop_interface?(shop_storage)
       end
 
       def user_storage=(storage)
         @user_storage = storage
+        return unless storage
```

Comment:
Same as above.

Reply:
We are using a null object pattern to support nil cases and if @user_storage is reassigned to nil, then it wouldn't actually set the value of that. The setter should ideally set the logic to whatever is passed in and handling of nil cases are handled through the pattern.